// position-map-text-to-markdown/lib/position-map-text-to-markdown.js
// MIT © 2017 azu
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var StringSource = require("textlint-util-to-string");
var parse = require("@textlint/markdown-to-ast").parse;
var TextSource = require("structured-source");
function getAllIndexes(matchText, allText) {
    var indexes = [];
    var i = -1;
    while ((i = allText.indexOf(matchText, i + 1)) !== -1) {
        indexes.push(i);
    }
    return indexes;
}
/**
 * get markdown location info that match `text`
 * @param {string} text
 * @param {string} markdown
 * @returns {{
 *  markdown:string,
 *  range: [number, number],
 *  loc: {
 *    start: {
 *       line:number,
 *       column:number
 *    },
 *    end: {
 *       line:number,
 *       column:number
 *    }
 * }
 * }[]}
 */
function findAllPositions(_a) {
    var text = _a.text, markdown = _a.markdown;
    var markdownAST = parse(markdown);
    var markdownRawSource = new TextSource(markdown);
    var markdownSource = new StringSource(markdownAST);
    var markdownText = markdownSource.toString();
    var matchIndexes = getAllIndexes(text, markdownText);
    // const parentMatchIndex = markdownText.indexOf(parentText);
    if (matchIndexes.length === 0) {
        return []; // no match
    }
    // get original info
    var getOriginalInfo = function (index, length) {
        var startIndex = markdownSource.originalIndexFromIndex(index);
        var endIndex = markdownSource.originalIndexFromIndex(index + length, true);
        var startPosition = markdownRawSource.indexToPosition(startIndex);
        var endPosition = markdownRawSource.indexToPosition(endIndex);
        return {
            markdown: markdown.slice(startIndex, endIndex),
            range: [startIndex, endIndex],
            loc: {
                start: startPosition,
                end: endPosition
            }
        };
    };
    return matchIndexes.map(function (matchIndex) {
        return getOriginalInfo(matchIndex, text.length);
    });
}
exports.findAllPositions = findAllPositions;
